sirepo/package_data/static/js/sirepo-components.js

```diff
@@ -4713,78 +4782,92 @@ SIREPO.app.directive('sbatchLoginModal', function() {
-                sbatchLoginStatusService.loginSuccess();
-                onHidden(r);
-                onHidden = null;
-                errorResponse = null;
-                $scope.otp = '';
-                $scope.password = '';
-                $scope.username = '';
-                $scope.sbatchLoginModalForm.$setPristine();
-                $scope.$apply();
-            });
-
-            function handleResponse(data) {
-                if (data.state === 'error') {
-                    errorResponse = data.error;
-                }
-                sbatchLoginStatusService.loggedIn = data.loginSuccess ? true: false;
-                $rootScope.$broadcast('sbatchLoginDone');
+        controller: function($element, $scope, $rootScope, authState, errorService, panelState, requestSender) {
+	    resetLoginFormTextFields();
+            let awaitingLoginResponse = false;
+	    let loginSuccess = false;
+	    let restartRunSimulation = false;
+            const el = $('#sbatch-login-modal');
+	    $scope.sbatchHostDisplayName = authState.sbatchHostDisplayName;
+
+            function handleLoginResponse(response) {
+		if (response.isSRException && response.data.routeName.toLowerCase().includes('sbatch')) {
+		    // POSIT: showing sbatch login modal is handled by requestSender.handleSRException
+		    return;
+		}
+		// This shouldn't happen
+		if (! response.loginSuccess) {
+		    srlog('Error with sbatchLogin response=', response);
+		    errorService.alertText('Error logging in. Please try again.');
+		    return;
+		}
+		loginSuccess = true;
                 el.modal('hide');
             }
 
-            $scope.$on('showSbatchLoginModal', function(e, broadcastArg) {
-                // When a user enters invalid login creds 'showSbatchLoginModal' is
-                // broadcast again. onHidden keeps a references to the
-                // errorCallback of only the first broadcast's errorCallback
-                const p = broadcastArg.srExceptionParams;
-                if (onHidden === null) {
-                    onHidden = broadcastArg.errorCallback;
-                }
-                sbatchLoginStatusService.loggedIn = false;
-                $scope.otp = '';
-                $scope.password = '';
-                awaitingSendResponse = false;
-                $scope.host = p.host;
-                $scope.showOtp = p.host.indexOf('nersc') >= 0;
-                $scope.showWarning = p.reason === 'invalid-creds';
-                $scope.warningText = 'Your credentials were invalid. Please try again.';
+	    function resetLoginForm() {
+		resetLoginFormTextFields();
+		$scope.sbatchLoginModalForm.$setPristine();
+                awaitingLoginResponse = false;
+	    }
+
+	    function resetLoginFormTextFields() {
+		$scope.otp = '';
+		$scope.password = '';
+		$scope.username = '';
+	    }
+
+            $scope.$on('showSbatchLoginModal', function(event, broadcastArg) {
+		restartRunSimulation = broadcastArg.shouldRestartRunSimulation;
```

Comment:
I think there would be better coupling if `addRestartRunSimulation()` (see below) looked at `srExceptionParams` (new name for broadcastArg) directly and the assignment wasn't here. 

In looking at this code, there are too many booleans, and this should be a state machine handled central. restartRunSimulation is both global and a parameter passed around in the same controller. This is very confusing.

Reply:
I very begrudgingly introduced this bit of state. If you have a way to eliminate it please let me know. But, I can't think of a way to get rid of it. The problem is that the shouldRestartSimulation come through in the broadcast to showSbatchLoginModal. Somehow we need to communicate shouldRestartRunSimulation to `el.on('hidden.bs.modal'...)`. The restartRunSimulation var was the only way I could figure out how to do this.